framework/include/base/MooseFunctor.h

```diff
@@ -352,48 +113,49 @@ class FunctorBase
   /**
    * Returns whether the functor is defined on this block
    */
-  virtual bool hasBlocks(const SubdomainID & /* id */) const
+  virtual bool hasBlocks(SubdomainID /* id */) const
   {
     mooseError("Block restriction has not been implemented for functor " + functorName());
     return false;
   }
 
   /**
-   * Returns a pair where the first member is whether this face is an extrapolated boundary face for
-   * this functor. The second member is the element on which this functor is defined if this is an
-   * extrapolated boundary face (if it is not an extrapolated boundary face, then we just return the
-   * face information \p &elem())
+   * Returns whether this (sided) face is an extrapolated boundary face for
+   * this functor
    */
-  virtual std::pair<bool, const Elem *> isExtrapolatedBoundaryFace(const FaceInfo &) const
+  virtual bool isExtrapolatedBoundaryFace(const FaceInfo &, const Elem *) const
   {
     mooseError("not implemented");
   }
 
   /**
    * Returns true if the face is an internal face
    */
-  virtual bool isInternalFace(const FaceInfo &) const { mooseError("not implemented"); }
+  bool isInternalFace(const FaceInfo &) const;
 
   /**
    * Returns true if this functor is a constant
    */
   virtual bool isConstant() const { return false; }
 
+  bool hasFaceSide(const FaceInfo & fi, const bool fi_elem_side) const override;
+
+  /**
+   * Check whether this functor is discontinuous at the face
+   * @param The face to check for discontinuity
+   * @return a tuple where the zeroth member indicates whether there is a discontinuity, the first
+   * member is the functor value on the "element" side of the face, and the second member is the
+   * functor value on the "neighbor" side of the face
+   */
+  std::tuple<bool, T, T> isDiscontinuous(const FaceInfo & fi) const;
```

Comment:
We do the evaluations in order to determine whether it is discontinuous. In the vein of the wasteful Dirichlet methods I referenced in the new issue, I don't like to throw away the work we've already done

Reply:
Is this just there to check if the porosity is discontinuous? Maybe we could have a virtual member for `isDiscontinuous` which could be overridden by the Porosity variable on the interfaces which we can cache whenever the mesh/porosity changes. 
For example, we can check on block boundaries if the porosity is different on the two sides or something simple. 
 I feel like we are trying to be super general here but the issue this PR wants to address is very specific. I'll think about this, this might be microoptimization.

Reply:
I could potentially revert the two `isDiscontinuous` commits. I used to have a `NS::isPorosityJumpFace` before I added the two `isDiscontinuous` commits. But the `NS` method returned the same kind of tuple. I think it makes sense to only check for porosity discontinuities on subdomain boundaries... I had a similar conversation with @snschune. That would be a nice optimization as opposed to checking for porosity discontinuities on every face. But I would still want the tuple return type when checking for a porosity jump at subdomain boundaries because why throw away the porosity evaluations we did to perform the check? Caching is always an option but we have only seen one instance where caching saves us time (the cell center gradients) and it adds code complexity.

Note that porosity functor could be an aux variable, a parsed function with `if` conditions in it, or a `PiecewiseByBlockLambdaFunctor` so porosity jump caching information would have to be handled in all of those cases.

Reply:
Got it, thanks for addressing this!